specifyweb/specify/migrations/0002_geo.py

```diff
@@ -0,0 +1,330 @@
+                ('collection', models.ForeignKey(db_column='CollectionID', on_delete=protect_with_blockers, related_name='cotypes', to='specify.collection')),
+                ('createdbyagent', models.ForeignKey(db_column='CreatedByAgentID', null=True, on_delete=protect_with_blockers, related_name='+', to='specify.agent')),
+                ('modifiedbyagent', models.ForeignKey(db_column='ModifiedByAgentID', null=True, on_delete=protect_with_blockers, related_name='+', to='specify.agent')),
+                ('taxontreedef', models.ForeignKey(db_column='TaxonTreeDefID', on_delete=protect_with_blockers, related_name='cotypes', to='specify.taxontreedef')),
+            ],
+            options={
+                'db_table': 'collectionobjecttype',
+                'ordering': (),
+            },
+        ),
+        migrations.CreateModel(
+            name='CollectionObjectGroupType',
+            fields=[
+                ('id', models.AutoField(db_column='COGTypeID', primary_key=True, serialize=False)),
+                ('name', models.CharField(db_column='Name', max_length=255, null=False)),
+                ('type', models.CharField(blank=True, db_column='Type', max_length=255, null=False)),
+                ('version', models.IntegerField(blank=True, db_column='Version', default=0, null=True)),
+                ('timestampcreated', models.DateTimeField(db_column='TimestampCreated', default=django.utils.timezone.now)),
+                ('timestampmodified', models.DateTimeField(blank=True, db_column='TimestampModified', default=django.utils.timezone.now, null=True)),
+                ('collection', models.ForeignKey(db_column='CollectionID', on_delete=protect_with_blockers, related_name='cogtypes', to='specify.collection')),
+                ('createdbyagent', models.ForeignKey(db_column='CreatedByAgentID', null=True, on_delete=protect_with_blockers, related_name='+', to='specify.agent')),
+                ('modifiedbyagent', models.ForeignKey(db_column='ModifiedByAgentID', null=True, on_delete=protect_with_blockers, related_name='+', to='specify.agent')),
+            ],
+            options={
+                'db_table': 'collectionobjectgrouptype',
+                'ordering': (),
+            },
+        ),
+        migrations.AddField(
+            model_name='collectionobject',
+            name='collectionobjecttype',
+            field=models.ForeignKey(db_column='CollectionObjectTypeID', null=True, on_delete=models.SET_NULL, related_name='collectionobjects', to='specify.collectionobjecttype'),
+        ),
+        migrations.AddField(
+            model_name='collection',
+            name='collectionobjecttype',
+            field=models.ForeignKey(db_column='CollectionObjectTypeID', null=True, on_delete=models.SET_NULL, related_name='collections', to='specify.collectionobjecttype'),
+        ),
+        migrations.RunPython(create_default_collection_types, revert_default_collection_types, atomic=True),
```

Comment:
Here's a blunder you are making (which i hoped it would have been caught). MySQL or MariaDB doesn't support DDL with transactions (they are implicit commits). You create a table, you do a commit implicitly. You make a filed change, you commit. Now, whatever sideeffect this function has, it'll be commiteed the moment the next operation runs. And if you encounter ANY error downstream, this will be committed. Move all non-DDLs to end, run them inside a transaction, rollback if they fail.